### ISSUE TITLE
Create organizations CLI

### ISSUE THREAD

**Issue by bmicklea** [codenvy/codenvy#636]
We need a CLI that will mirror the API and allow users to use CLI syntax to make changes to the organization.

The CLI commands should be simple, clear and human readable. Feel free to outline commands here for us to discuss.

#### Update:
List of Organization CLI commands:
- [ ] Create organization
_POST_
`codenvy organization create --name={name} --parent={parent-id}`

- [ ] Update organization 
_POST /{id}_
`codenvy organization update --id={id} --name={name} --new-name={new-name}`

- [ ] Delete organization 
_DELETE /{id}_
`codenvy organization delete --id={id} --name={name}`

- [ ] Review organization 
_GET /{organizationId}_, _GET /find?name={name}_
`codenvy organization --id={id} --name={name}`
> {
>    "id": "{id}"
>    "name": "{name}"
>    "parent": "{parent-id}"
> }

- [ ] Get list of children: 
_GET /{parent}/organizations_
`codenvy organization children --id={id} --name={name}`
> [{
>    "id": "{id1}"
>    "name": "{name1}"
>    "parent": "{parent-id1}"
> },
> {
>    "id": "{id2}"
>    "name": "{name2}"
>    "parent": "{parent-id2}"
> },
> ..]

- [ ] Get list of organizations where user is member (#669)
_GET ?user={user}_ 
`codenvy organization list --user={user}`
> [{
>    "id": "{id1}"
>    "name": "{name1}"
>    "parent": "{parent-id1}"
> },
> {
>    "id": "{id2}"
>    "name": "{name2}"
>    "parent": "{parent-id2}"
> },
> ..]

- [ ] User should login using "*codenvy login*" command before executing *organization* command. it is needed to get access to _Organization API_. To do this, next issue should be resolved (#667).

This issue is blocked by the next Issues which should be fixed before:
- [ ] https://github.com/codenvy/codenvy/issues/410
- [ ] https://github.com/codenvy/codenvy/issues/419
- [ ] https://github.com/codenvy/codenvy/issues/669

#### Resources
- [Codenvy 4 Account for resources management](https://wiki.codenvycorp.com/pages/viewpage.action?title=Codenvy+4+Account+for+resources+management&spaceKey=PSR)
- [OrganizationService](https://github.com/sleshchenko/codenvy/blob/5c2c4b0bfd8f125c5583a6c8cd6c027cbdd48ebf/wsmaster/codenvy-hosted-api-organization/src/main/java/com/codenvy/organization/api/OrganizationService.java
)
- [OrganizationManager](https://github.com/sleshchenko/codenvy/blob/5c2c4b0bfd8f125c5583a6c8cd6c027cbdd48ebf/wsmaster/codenvy-hosted-api-organization/src/main/java/com/codenvy/organization/api/OrganizationManager.java
)
- [OrganizationDto](https://github.com/sleshchenko/codenvy/blob/5c2c4b0bfd8f125c5583a6c8cd6c027cbdd48ebf/wsmaster/codenvy-hosted-api-organization/src/main/java/com/codenvy/organization/shared/dto/OrganizationDto.java
)

**Comment by bmicklea** [comment]
@dmytro-ndp I had to move this issue so you'd made the following comment:
>It seems we should have this issue be done before https://github.com/codenvy/codenvy/issues/552.

>But Organization REST API is now strictly coupled to JPA Integration: https://github.com/codenvy/codenvy/issues/409#issuecomment-235884484

**Comment by tolusha** [comment]
This issue depends on organization stuff that isn't at master  yet. Organization corespondently depends on jpa. That's why we can't guarantee to finish this issue in time. At least we can't start it in dedicated branch.

**Comment by tolusha** [comment]
@skabashnyuk Could you pls confirm dependencies above ?

**Comment by TylerJewell** [comment]
@tolusha - those dependencies make sense to me.

**Comment by tolusha** [comment]
[Account for resources management](https://wiki.codenvycorp.com/pages/viewpage.action?title=Codenvy+4+Account+for+resources+management&spaceKey=PSR)

**Comment by dmytro-ndp** [comment]
Another dependencies: https://github.com/codenvy/codenvy/issues/419

**Comment by dmytro-ndp** [comment]
@benoitf, @sleshchenko: please, take a look on my proposal based on the content of [Organization API implementation](https://github.com/sleshchenko/codenvy/blob/5c2c4b0bfd8f125c5583a6c8cd6c027cbdd48ebf/wsmaster/codenvy-hosted-api-organization/src/main/java/com/codenvy/organization/api/OrganizationService.java):

#### Create organization command:
`codenvy organization create --user={user} --password={password} --name={name} --parent={parent-id}
`
####  Update organization command:
`codenvy organization update --user={user} --password={password} --id={id} --name={name} --new-name={new-name} --new-parent={parent-id}`

#### Remove organization command:
`codenvy organization remove --user={user} --password={password} --id={id} --name={name}`

#### Review organization command:
`codenvy organization --user={user} --password={password} --id={id} –name={name}`
> {
>    "id": "{id}"
>    "name": "{name}"
>    "parent": "{parent-id}"
>    "childrens": ["children1-id", "children2-id", …]
> }

#### Get list of organizations where user is member command:
`codenvy organizations --user={user} --password={password}`
> [{
>    "id": "{id1}"
>    "name": "{name1}"
>    "parent": "{parent-id1}"
>    "childrens": ["children1.1-id", "children1.2-id", …]
> },
> {
>    "id": "{id2}"
>    "name": "{name2}"
>    "parent": "{parent-id2}"
>    "childrens": ["children2.1-id", "children2.2-id", …]
> },
> ..]

#### Resources
- [Codenvy 4 Account for resources management](https://wiki.codenvycorp.com/pages/viewpage.action?title=Codenvy+4+Account+for+resources+management&spaceKey=PSR)
- [OrganizationService](https://github.com/sleshchenko/codenvy/blob/5c2c4b0bfd8f125c5583a6c8cd6c027cbdd48ebf/wsmaster/codenvy-hosted-api-organization/src/main/java/com/codenvy/organization/api/OrganizationService.java
)
- [OrganizationManager](https://github.com/sleshchenko/codenvy/blob/5c2c4b0bfd8f125c5583a6c8cd6c027cbdd48ebf/wsmaster/codenvy-hosted-api-organization/src/main/java/com/codenvy/organization/api/OrganizationManager.java
)
- [OrganizationDto](https://github.com/sleshchenko/codenvy/blob/5c2c4b0bfd8f125c5583a6c8cd6c027cbdd48ebf/wsmaster/codenvy-hosted-api-organization/src/main/java/com/codenvy/organization/shared/dto/OrganizationDto.java
)

**Comment by TylerJewell** [comment]
When you are creating an organization, can you comment a little bit on why there is a --user and --password?  Does this mean that each organization is "owned" by a single user / password combination?

**Comment by bmicklea** [comment]
For listing organizations for a user I think a more logical syntax would be:

`codenvy list-organizations --user={user} --password={password}`

**Comment by dmytro-ndp** [comment]
@TylerJewell: 
> When you are creating an organization, can you comment a little bit on why there is a --user and --password? Does this mean that each organization is "owned" by a single user / password combination?

--user and --password are needed for user authorization. They also uses to get list of organization where user is a member.

**Comment by dmytro-ndp** [comment]
@bmicklea:
> For listing organizations for a user I think a more logical syntax would be:
> codenvy list-organizations --user={user} --password={password}

I tried to avoid using "-" in command name for simplicity.
We could use next command instead:
`codenvy organization list --user={user} --password={password}`

**Comment by TylerJewell** [comment]
Are those parameters optional?  I thought there was a Codenvy login cli entry that would give you a token for the session.

**Comment by dmytro-ndp** [comment]
Login command exists but it is aimed for login to remote Codenvy cloud, that is codenvy.com by default.

**Comment by TylerJewell** [comment]
Can you elaborate?  I thought the point of the login command is to get a local token for any Codenvy system?

Then the user and password entries would be optional and we would document them as such.  It is not good practice to have the user pass in entries every invocation.

**Comment by dmytro-ndp** [comment]
We could rewrite login command to use it for login to local Codenvy on-prem instance. It allows to avoid using --user/--password parameters.

**Comment by TylerJewell** [comment]
Are you sure it doesn't already work that way? I remember designing it two years ago to act that way - giving user a reusable token that is applied for all other cli commands.  :)

**Comment by dmytro-ndp** [comment]
It was built on the top of Codenvy CLI login command which is used to preform operations with Codenvy 3.x SaaS.
IM CLI used login to get Codenvy SaaS subscription. 
Because of this, default remote address of Codenvy is still "codenvy.com".

**Comment by tolusha** [comment]
What about to rename `--user` -> `--adminUser` and `--password` -> `--adminPassword` ? It will be more clear for user.

**Comment by dmytro-ndp** [comment]
@tolusha: yes, you are right. The only moment is getting list of organizations for user which is not admin.

**Comment by TylerJewell** [comment]
-1. I just don't know of any cli that requires user to odd their auth credentials on every invocation.

**Comment by tolusha** [comment]
We already decided to reimplement login  command to be able to login into installed onprem.
So, no user credentials will be required.

**Comment by TylerJewell** [comment]
Ok.  I didn't see the additional post - github on phone had a cache issue.  Let's update the syntax and maybe we should delete this long thread as it may no longer be useful for others to see our back and forth.

**Comment by tolusha** [comment]
I mean it was our internal office conversion. @dmytro-ndp will create the issue.

**Comment by dmytro-ndp** [comment]
@TylerJewell: that is the issue to fix IM CLI *login* command:
https://github.com/codenvy/codenvy/issues/667

This fix could be helpful for "*codenvy audit*" command as well.

**Comment by tolusha** [comment]
It is not actual anymore. We have taken a decision to add organization management into Dashboard and to document how to manage organization using curl operations.

**Comment by vkuznyetsov** [comment]
Won't fix. Closed.

**Comment by bmicklea** [comment]
@tolusha is there an issue to describe the dashboard work?
